drivers/platform/maxim/maxim_uart.c

```diff
@@ -0,0 +1,353 @@
+
+	int32_t err = 0;
+	uint32_t bytes_read = 0;
+
+	int32_t len = bytes_number;
+	err = UART_Read(MXC_UART_GET_UART(desc->device_id), data, len, &bytes_read);
+
+	if(err < 0)
+		return -EIO;
+
+	return bytes_read;
+}
+
+/**
+ * @brief Write data to UART device. Blocking function.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to read.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t uart_write(struct uart_desc *desc, const uint8_t *data,
+		   uint32_t bytes_number)
+{
+	if(!desc || !data || !bytes_number)
+		return -EINVAL;
+
+	uint32_t err = 0;
+
+	int32_t len = bytes_number;
+	err = UART_Write(MXC_UART_GET_UART(desc->device_id), data, len);
+
+	if(err < 0)
+		return -EIO;
+
+	return 0;
+}
+
+/**
+ * @brief Read data from UART device. Non blocking function.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to read.
+ * @return positive number of received bytes in case of success, negative error code otherwise.
+ */
+int32_t uart_read_nonblocking(struct uart_desc *desc, uint8_t *data,
+			      uint32_t bytes_number)
+{
+	if(!desc || !data || !bytes_number)
+		return -EINVAL;
+
+	uint32_t err = 0;
+	uart_req_t read_req;
+
+	read_req.data = data;
+	read_req.len = bytes_number;
+	read_req.callback = NULL;
+	err = UART_ReadAsync(MXC_UART_GET_UART(desc->device_id), &read_req);
+
+	return err;
+}
+
+/**
+ * @brief Write data to UART device. Non blocking function.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to read.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+
+int32_t uart_write_nonblocking(struct uart_desc *desc, const uint8_t *data,
+			       uint32_t bytes_number)
+{
+	if(!desc || !data || !bytes_number)
+		return -EINVAL;
+
+	uint32_t err = 0;
+	uart_req_t write_req;
+
+	write_req.data = data;
+	write_req.len = bytes_number;
+	write_req.callback = NULL;
+	err = UART_WriteAsync(MXC_UART_GET_UART(desc->device_id), &write_req);
+
+	return err;
+}
+
+/**
+ * @brief Initialize the UART communication peripheral.
+ * @param desc - The UART descriptor.
+ * @param param - The structure that contains the UART parameters.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t uart_init(struct uart_desc **desc, struct uart_init_param *param)
+{
+	if(!param || !desc)
+		return -EINVAL;
+	int ret = 0;
+	uint32_t did = param->device_id;
+
+	struct maxim_uart_desc *maxim_uart = calloc(1, sizeof(*maxim_uart));
+	struct uart_desc *descriptor = calloc(1, sizeof(*descriptor));
+
+	uart_cfg_t maxim_desc;
+	sys_cfg_uart_t maxim_desc_sys;
+
+	if(!maxim_uart || !descriptor) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	NVIC_ClearPendingIRQ(MXC_UART_GET_IRQ(did));
+	NVIC_DisableIRQ(MXC_UART_GET_IRQ(did));
+	NVIC_SetPriority(MXC_UART_GET_IRQ(did), 1);
```

Comment:
Question: why are we setting interrupt priority of 1 here? And is there a way where we could do this in an IRQ driver? Status quo right now is to [include the IRQ driver](https://github.com/analogdevicesinc/no-OS/blob/master/drivers/platform/xilinx/uart.c#L42) in the other ones to access IRQ specific functionality.